packages/lesswrong/server/resolvers/ultraFeedResolver.ts

```diff
@@ -169,118 +138,97 @@ const createSourcesMap = (
-    }
-
-    sources[sourceType] = {
-      weight,
-      items: [], 
-      renderAsType
-    };
+  spotlightItems: FeedSpotlight[],
+  bookmarkItems: PreparedBookmarkItem[]
+): Partial<Record<FeedItemSourceType, WeightedSource>> => {
+  const sources: Partial<Record<FeedItemSourceType, WeightedSource>> = {};
+  Object.entries(sourceWeights).forEach(([src, w]) => {
+    if (w > 0) sources[src as FeedItemSourceType] = { weight: w, items: [] };
   });
 
-  if (sources.spotlights && spotlightItems.length > 0) {
-    sources.spotlights.items = spotlightItems;
-  }
+  const push = (src: FeedItemSourceType, itm: SampledItem) => {
+    const bucket = sources[src];
+    if (!bucket) return;
 
-  postThreadsItems.forEach(postItem => {
-    const itemSources = postItem.postMetaInfo?.sources;
-    if (Array.isArray(itemSources)) {
-      itemSources.forEach(source => {
-        const sourceType = source as FeedItemSourceType;
-        if (sources[sourceType]) {
-          // Avoid adding duplicates to the same source list if a post somehow has multiple identical sources
-          if (!sources[sourceType].items.some(item => (item as FeedFullPost).post?._id === postItem.post?._id)) {
-             sources[sourceType].items.push(postItem);
-          }
-        }
-      });
+    let itemId: string | undefined;
+    if (itm.type === "feedPostWithContents") {
+      itemId = itm.feedPost?.post?._id;
+    } else if (itm.type === "feedPost") {
+      itemId = itm.feedPostStub.postId;
     }
-  });
 
-  // Add ALL prepared comment threads to the single 'recentComments' bucket if it exists
-  if (sources[commentSourceType]) {
-    sources[commentSourceType].items = commentThreadsItems; 
-  } else if (commentThreadsItems.length > 0) {
-      // eslint-disable-next-line no-console
-      console.warn(`UltraFeedResolver: Prepared ${commentThreadsItems.length} comment threads, but '${commentSourceType}' source is not defined or has zero weight in sourceWeights.`);
-  }
+    if (itemId && bucket.items.some(existing => {
+      if (existing.type === "feedPostWithContents") return existing.feedPost?.post?._id === itemId;
+      if (existing.type === "feedPost") return existing.feedPostStub.postId === itemId;
+      return false;
+    })) return;
 
-  return Object.entries(sources).reduce((acc, [key, value]) => {
-    if (value.items.length > 0) {
-      acc[key as FeedItemSourceType] = value;
-    }
-    return acc;
-  }, {} as Record<FeedItemSourceType, WeightedSource>);
+    bucket.items.push(itm);
+  };
+
+  spotlightItems.forEach(s => push("spotlights", { type: "feedSpotlight", feedSpotlight: s }));
+
+  postThreadsItems.forEach(p =>
+    (p.postMetaInfo?.sources ?? []).forEach(src =>
+      push(src as FeedItemSourceType, { type: "feedPostWithContents", feedPost: p })));
+
+  commentThreadsItems.forEach(t =>
+    push("recentComments", { type: "feedCommentThread", feedCommentThread: t }));
+
+  bookmarkItems.forEach(item => push("bookmarks", item));
```

Comment:
Each of these 4 should be something that returns a single object, which you then return as part of a single object:
```
const spotlightSources = ...;
const postThreadSources = ...;
...

return {
  spotlights: spotlightSources,
  feedPostWithContents: postThreadSources,
  ...
};
```